examples/gazebo_differential_drive_plugin/differential_drive_plugin.cc

```diff
@@ -0,0 +1,110 @@
+    /// attached to.
+    /// \param[in] _sdf A pointer to the plugin's SDF element.
+    public: virtual void Load(physics::ModelPtr _model, sdf::ElementPtr _sdf)
+    {
+      // Safety check
+      if (_model->GetJointCount() == 0)
+      {
+        std::cerr << "Invalid joint count, DifferentialDrivePlugin not loaded\n";
+        return;
+      }
+
+      // Store the model pointer for convenience.
+      this->model = _model;
+
+      // Get the first 2 joint. We are making an assumption about the model
+      // having 2 joints which are the 2 wheels.
+      this->left_wheel_joint = _model->GetJoints()[1];
+      this->right_wheel_joint = _model->GetJoints()[0];
+
+      // Setup a P-controller, with a gain of 0.1.
+      this->left_wheel_pid = common::PID(10, 0, 0);
+      this->right_wheel_pid = common::PID(10, 0, 0);
+
+      // Apply the P-controller to the joint.
+      this->model->GetJointController()->SetVelocityPID(this->left_wheel_joint->GetScopedName(), this->left_wheel_pid);
+      this->model->GetJointController()->SetVelocityPID(this->right_wheel_joint->GetScopedName(), this->right_wheel_pid);
+
+      // Default to zero velocity
+      double velocity = 0;
+
+      // Check that the velocity element exists, then read the value
+      if (_sdf->HasElement("velocity"))
+        velocity = _sdf->Get<double>("velocity");
+
+      this->SetVelocity(velocity, velocity);
+
+      // Create the node
+      this->node = transport::NodePtr(new transport::Node());
+      #if GAZEBO_MAJOR_VERSION < 8
+      this->node->Init(this->model->GetWorld()->GetName());
+      #else
+      this->node->Init(this->model->GetWorld()->Name());
+      #endif
+
+      // Create a topic name
+      std::string topicName = "~/" + this->model->GetName() + "/vel_cmd";
+
+      // Subscribe to the topic, and register a callback
+      this->sub = this->node->Subscribe(topicName, &DifferentialDrivePlugin::OnMsg, this);
+    }
+
+    /// \brief Set the velocity of the Velodyne
+    /// \param[in] _vel New target velocity
+    public: void SetVelocity(const double &_left_wheel_vel, const double &_right_wheel_vel)
+    {
+      // Set the joint's target velocity.
+      this->model->GetJointController()->SetVelocityTarget(this->left_wheel_joint->GetScopedName(), -(_left_wheel_vel));
+      this->model->GetJointController()->SetVelocityTarget(this->right_wheel_joint->GetScopedName(), _right_wheel_vel);
+      // std::cerr<< this->left_wheel_joint->GetScopedName()<< ": " <<_left_wheel_vel << "." << this->right_wheel_joint->GetScopedName() << ": "<< _right_wheel_vel << std::endl;
+    }
+
+    /// \brief Handle incoming message
+    /// \param[in] _msg Repurpose a vector3 message. This function will
+    /// only use the x component.
+    private: void OnMsg(ConstVector3dPtr &_msg)
+    {
+      this->SetVelocity(_msg->x(), _msg->y());
+    }
+
+    /// \brief A node used for transport
+    private: transport::NodePtr node;
```

Comment:
I don't *think* this needs to be a pointer. I think you could just write this as:
```
transport::Node node;
```

And then it'll be created automatically and you won't have to do this above:
```c++
this->node = transport::NodePtr(new transport::Node());
```

Reply:
This was modified from the Gazebo tutorial [Intermediate: Control plugin](http://gazebosim.org/tutorials?tut=guided_i5). I tried to set it to `transport::Node node;`, but the plugin crashed. It seems like the transport node is a shared pointer which gets passed to Gazebo's transport system when we call `node->Subscribe`. This is how it's defined in `gazebo/transport/TransportTypes.hh`:
```cpp
/// \def NodePtr
/// \brief Shared_ptr to Node object
typedef boost::shared_ptr<Node> NodePtr;
```
I think the `DifferentialDrivePlugin` object gets destroyed after `Load`, so when I tried to implement with just `transport::Node node`, the node object also got deleted leading to a crash.

Reply:
Fair enough.